### PR TITLE
[SPARK-40756][SQL] Migrate type check failures of string expressions onto error classes

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
This pr replace TypeCheckFailure by DataTypeMismatch in type checks in the string expressions, includes:
- regexpExpressions.scala (RegExpReplace)
- stringExpressions.scala (Etl)

### Why are the changes needed?
Migration onto error classes unifies Spark SQL error messages.

### Does this PR introduce _any_ user-facing change?
Yes. The PR changes user-facing error messages.

### How was this patch tested?
- Add new UT
- Update existed UT
- Pass GA.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @MaxGekk

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -248,11 +248,21 @@
           "The <exprName> must not be null"
         ]
       },
+      "UNEXPECTED_NUM_PARAMS" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:251]: It seems it is similar to `WRONG_NUM_PARAMS`. Could you re-use the existing one or improve it.

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -248,11 +248,21 @@
...
+        "message" : [
+          "The <functionName> requires at least <num> arguments."
+        ]
+      },
       "UNEXPECTED_STATIC_METHOD" : {
         "message" : [
           "cannot find a static method <methodName> that matches the argument types in <className>"
         ]
       },
+      "UNEXPECTED_VALUE" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:261]: Can't you re-use `VALUE_OUT_OF_RANGE`?

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (LuciferYang) [reply core/src/main/resources/error/error-classes.json:251]: I think we can make `WRONG_NUM_PARAMS` more general

Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json:251]: Can I change it to：
"WRONG_NUM_PARAMS" : {
        "message" : [
          "The <functionName> requires <num> arguments (actual number = <actualNum>)"
        ]
      }

File: core/src/main/resources/error/error-classes.json (hunk 2 above)
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json]: Done

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json]: Please, open a separate PR and change the error template. I would change it to:
```json
      "WRONG_NUM_PARAMS" : {
        "message" : [
          "The <functionName> requires <expectedNum> parameters but the actual number is <actualNum>."
        ]
      }
```

Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json]: ok

**Comment by panbingkun** [comment]
After https://github.com/apache/spark/pull/38319, I will continue do it.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun and @LuciferYang for review.

**Pull request closed without merging**